### PR TITLE
feat: inference_fn changed from asynchronous to synchronous

### PR DESCRIPTION
**Pull request by lazyky**
- fix bugs
- complete function test
- delete the directory parameter in TaskConfig

Resove #1

### CODE REVIEW

File: cml-core/src/core/inference.rs [hunk 1]
```diff
@@ -15,15 +15,15 @@ pub struct NewSample<F> {
...
+pub trait Inference<M, F, T, C: Manager> {
     async fn init_inference(
         &self,
         target_type: M,
         optional_fields: Option<Vec<M>>,
         optional_tags: Option<Vec<M>>,
     ) -> Result<()>;
 
-    async fn inference<FN, R>(
+    async fn inference<FN>(
```
Comment (bioinformatist) [inline cml-core/src/core/inference.rs:26]: I think here we need a parameter `available_status`, just like `working_status` in `task.rs`, which allows user to custom the list of available status. For example, one could mark some status for a specific type of inference, but use others for another inference operation.

File: cml-tdengine/src/core/inference.rs [hunk 2]
```diff
@@ -48,15 +46,17 @@ impl<D: IntoDsn + Clone> Inference<Field, Value, Manager<TaosBuilder>> for TDeng
...
     {
-        let samples_with_res = inference_fn(data, pool).await?;
-
         let taos = pool.get().await?;
-        taos.use_database("training_data").await?;
         let mut stmt = Stmt::init(&taos)?;
-
+        taos.use_database("task").await?;
+        let last_task_time = taos
+            .query_one(format!("SELECT LAST(ts) FROM task.`{}`", metadata.batch()))
```
Comment (bioinformatist) [inline cml-tdengine/src/core/inference.rs:55]: Here we need to pick the task according to `available_status`.

BTW, `query_one` is a new API which is a better option for these scenarios. Could you replace the original code with it in other appropriate places?

**Review by bioinformatist — CHANGES_REQUESTED** [review]

File: cml-core/src/core/inference.rs (hunk 1 above)
Comment (bioinformatist) [inline cml-core/src/core/inference.rs:26]: I think here we need a parameter available_status, just like working_status in task.rs, which allows user to custom the list of available status. For example, one could mark some status for a specific type of inference, but use others for another inference operation.

**Review by bioinformatist — CHANGES_REQUESTED** [review]
The design of the trait method `inference` should be improved to handle more flexible situations.

**Review by bioinformatist — CHANGES_REQUESTED** [review]
Go go go!

File: cml-tdengine/src/core/inference.rs [hunk 3]
```diff
@@ -149,6 +163,146 @@ mod tests {
 
     #[tokio::test(flavor = "multi_thread")]
     async fn test_concurrent_inference() -> Result<()> {
+        let cml = TDengine::from_dsn("taos://");
+        let taos = cml.build_sync()?;
```
Comment (bioinformatist) [inline cml-tdengine/src/core/inference.rs:167]: Why connection by blocking here?

File: cml-tdengine/src/core/inference.rs [hunk 4]
```diff
@@ -149,6 +163,146 @@ mod tests {
...
+                    available_status,
+                    &mut batch_data_1,
+                    &pool,
+                    inference_fn,
+                )
+                .await
+                .unwrap();
+            }
+        })
+        .await?;
```
Comment (bioinformatist) [inline cml-tdengine/src/core/inference.rs:294]: We may need more than one batch running simultaneously to validate if async works well.

File: cml-tdengine/src/core/register.rs [hunk 5]
```diff
@@ -250,11 +250,11 @@ mod tests {
...
-        let mut result = taos
-            .query("SELECT COUNT(*) FROM training_data.training_data")
-            .await?;
-        let records = result.to_records().await?;
-        assert_eq!(vec![vec![Value::BigInt(4)]], records);
+        let records = taos
+            .query_one("SELECT COUNT(*) FROM training_data.training_data")
+            .await?
+            .unwrap_or(0);
+        assert_eq!(4, records);
```
Comment (bioinformatist) [inline cml-tdengine/src/core/register.rs:257]: LGTM

**Review by bioinformatist — CHANGES_REQUESTED** [review]
Test code for `Inference` trait needs to be changed.

File: cml-tdengine/src/core/inference.rs (hunk 3 above)
Comment (lazyky) [reply cml-tdengine/src/core/inference.rs]: That was my mistake

File: cml-tdengine/src/core/inference.rs [hunk 6]
```diff
@@ -149,6 +170,204 @@ mod tests {
...
+                    model_inference,
+                );
+                result.push(
+                    NewSampleBuilder::default()
+                        .data_path(inference_data.data_path().to_path_buf())
+                        .output(output)
+                        .build()
+                        .unwrap(),
+                );
+            }
```
Comment (bioinformatist) [inline cml-tdengine/src/core/inference.rs:327]: There's no need for use a closure here, for the closure won't be used more than one time. Just move the calculations into loop, and use variables for staging intermediate results if necessary.

**Review by bioinformatist — CHANGES_REQUESTED** [review]
Almost done!

**Review by bioinformatist — APPROVED** [review]

**Pull request merged**